### ISSUE TITLE
autograd grad_outputs are zero

### ISSUE THREAD

**Issue by euhruska** [pytorch/pytorch#6976]
I want to take the gradient of a function (vector) by x for each of the x values. The purpose is that this gradient is part of the loss function.
I tried autograd.grad with grad_outputs, but it gives zero gradients (df_dx).  They should be non-zero, so either I made a mistake or this is a bug.
```
a0=torch.tensor(1).double().cuda().requires_grad_(True)
x=torch.randn(10).double().cuda().requires_grad_(True)
f=torch.tanh(x-a0).requires_grad_(True)
df_dx=x=torch.zeros(10).double().cuda().requires_grad_(True)
grad_params = torch.autograd.grad(f, [x], create_graph=True,grad_outputs=[df_dx],allow_unused=True)
```

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Red Hat Enterprise Linux Workstation release 7.4 (Maipo)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)
CMake version: version 2.8.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 387.26
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6.0.21
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] numpy (1.13.3)
[pip] numpydoc (0.7.0)
[pip] torch (0.4.0)
[pip] torchvision (0.2.1)
[conda] pytorch                   0.4.0           py36_cuda8.0.61_cudnn7.1.2_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch

**Comment by SsnL** [comment]
1. Your `grad_output` is `0`. So you are calculating `dl/df * df/dx`, where `dl/df = 0`. I don't see how the result should be non-zero. If you want to find jacobian, you need to do autograd for each output value.
2. you don't need to set `requires_grad` to `True` every where. In your case `f` already has this because both `a0` and `x` used to compute it `requries_grad`. And `df_dx` isn't involved in any computation, so it doesn't need either.
